client/virtual_channel.go

```diff
@@ -100,6 +100,52 @@ func (c *Client) handleVirtualChannelFundingProposal(
 	c.acceptProposal(responder)
 }
 
+func (c *Channel) watchVirtual() error {
+	log := c.Log().WithField("proc", fmt.Sprintf("virtual channel watcher %v", c.ID()))
+	defer log.Info("Watcher returned.")
+
+	// Subscribe to state changes
+	ctx := c.Ctx()
+	sub, err := c.adjudicator.Subscribe(ctx, c.Params())
+	if err != nil {
+		return errors.WithMessage(err, "subscribing to adjudicator state changes")
+	}
+	defer func() {
+		err := sub.Close()
+		if err != nil {
+			log.Warn(err)
+		}
+	}()
+
+	// Wait for state changed event
+	for e := sub.Next(); e != nil; e = sub.Next() {
+		log.Debugf("event %v", e)
+
+		// Update channel
+		switch e := e.(type) {
+		case *channel.RegisteredEvent:
+			if e.Version() > c.State().Version {
+				err := c.client.updateVirtualChannelState(ctx, e.State, e.Sigs)
+				if err != nil {
+					log.Warnf("error updating virtual channel: %v", err)
+				}
+			}
+
+		case *channel.ProgressedEvent:
+			log.Infof("Virtual channel progressed: %v", e.ID())
```

Comment:
Persist state?

Reply:
Having thought about this:

Enabling virtual app channels turns out more complex than anticipated.

observation:
- hub only receives actor signature
- need all signatures for registering state

problem:
- hub cannot register withdrawal of mirror channel
- cross-chain: state registered only on one chain, not registered on other chain

solution:
- option1 (low effort, restricts functionality)
  - do not allow virtual app channels
  - do not allow virtual channels to have app sub-channels
- option2 (medium effort, does not work for cross-chain)
  - allow to omitt signature if state already registered
  - don't worry about cross-chain for now
- option3 (high effort, general solution)
  - implement option2
  - persist all state progressions
  - replay state progression on other chain

Reply:
Going with option1 for now